workflow/gccontroller/gc_controller.go

```diff
@@ -82,6 +97,75 @@ func (c *Controller) runWorker() {
 	}
 }
 
+type count struct {
+	errored int
+	failed  int
+}
+
+// retentionGC queues workflows for deletion based upon the retention policy.
+func (c *Controller) retentionGC() {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
+	// Change the index here to the more flexible one instead.
+	var uns []*unstructured.Unstructured
+	var objs []interface{}
+	for _, phase := range []wfv1.NodePhase{wfv1.NodeSucceeded, wfv1.NodeFailed, wfv1.NodeError} {
```

Comment:
This performs an O(n) operation each time. I think it is possible to reduce this to O(1). The way to do this would be to add the workflow to an ordered queue whenever a workflow is done (done = completed but not pending archive).  You could use a heap for this (they can be re-ordered whenever an item is added).

Whenever an item is added to the queue, then it is possible that an item needs to be deleted. The queue is longer than the retention amount, in fact, you can repeatedly delete the head item until the queue is the correct length.

This means we don't need to run this loop at all.

Reply:
Ok, In that case I suggest that we skip the setting for` error`, `failed` and `completed` and count them all towards the allowed workflows to retain.

Reply:
I think you might need a queue for all three states.

Reply:
But in that case we will only count towards the total nbr of workflows to keep. We could also have three queues and allow separate retention, but maybe we should keep it to one for the first itteration?

Reply:
use as many queues as you need